// @canonical/react-components/dist/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useClickOutside", {
  enumerable: true,
  get: function get() {
    return _useClickOutside.useClickOutside;
  }
});
Object.defineProperty(exports, "useId", {
  enumerable: true,
  get: function get() {
    return _useId.useId;
  }
});
Object.defineProperty(exports, "useListener", {
  enumerable: true,
  get: function get() {
    return _useListener.useListener;
  }
});
Object.defineProperty(exports, "useOnEscapePressed", {
  enumerable: true,
  get: function get() {
    return _useOnEscapePressed.useOnEscapePressed;
  }
});
Object.defineProperty(exports, "usePagination", {
  enumerable: true,
  get: function get() {
    return _usePagination.usePagination;
  }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function get() {
    return _usePrevious.usePrevious;
  }
});
Object.defineProperty(exports, "useThrottle", {
  enumerable: true,
  get: function get() {
    return _useThrottle.useThrottle;
  }
});
Object.defineProperty(exports, "useWindowFitment", {
  enumerable: true,
  get: function get() {
    return _useWindowFitment.useWindowFitment;
  }
});

var _useClickOutside = require("./useClickOutside");

var _useId = require("./useId");

var _useListener = require("./useListener");

var _useOnEscapePressed = require("./useOnEscapePressed");

var _usePrevious = require("./usePrevious");

var _useThrottle = require("./useThrottle");

var _usePagination = require("./usePagination");

var _useWindowFitment = require("./useWindowFitment");
